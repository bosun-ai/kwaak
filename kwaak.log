OpenTelemetry tracing enabled
>> starting up agent for the first time, this might take a while
