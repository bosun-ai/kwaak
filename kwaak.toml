language = "rust"

tavily_api_key = "env:TAVILY_API_KEY"
github_api_key = "env:GITHUB_TOKEN"
openai_api_key = "env:KWAAK_OPENAI_API_KEY"

tool_executor = "docker"
otel_enabled = true

[commands]
test = "RUST_LOG=kwaak=debug,swiftide=debug RUST_BACKTRACE=1 cargo nextest run --no-fail-fast --color=never"
coverage = "cargo llvm-cov nextest --no-clean --summary-only"
lint_and_fix = "cargo clippy --fix --allow-dirty --allow-staged; cargo fmt"

[git]
owner = "bosun-ai"
repository = "kwaak"
main_branch = "master"

[llm.indexing]
provider = "OpenAI"
prompt_model = "gpt-4o-mini"

[llm.query]
provider = "OpenAI"
prompt_model = "gpt-4o"

[llm.embedding]
provider = "OpenAI"
embedding_model = "text-embedding-3-large"

[docker]
dockerfile = "Dockerfile"
