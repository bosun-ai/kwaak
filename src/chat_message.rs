/// Represents a chat message that can be stored in a [`Chat`]
///
/// Messages are expected to be formatted strings and are displayed as-is. Markdown is rendered
/// using `tui-markdown`.
///
/// TODO: All should be Cows
#[derive(Clone, Default, PartialEq)]
pub struct ChatMessage {
    role: ChatRole,
    formatted_content: String,
    original: Option<swiftide::chat_completion::ChatMessage>,
}

// Debug with truncated content
impl std::fmt::Debug for ChatMessage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ChatMessage")
            .field("role", &self.role)
            .field(
                "content",
                &self.formatted_content[..std::cmp::min(10, self.formatted_content.len())]
                    .to_string(),
            )
            .field("original", &self.original)
            .finish()
    }
}

#[derive(
    Debug,
    Clone,
    Copy,
    Default,
    strum::EnumString,
    strum::Display,
    strum::AsRefStr,
    strum::EnumIs,
    PartialEq,
)]
pub enum ChatRole {
    User,
    #[default]
    System,
    Command,
    Assistant,
    Tool,
}

impl ChatMessage {
    pub fn new_user(msg: impl Into<String>) -> ChatMessage {
        ChatMessage::default()
            .with_role(ChatRole::User)
            .with_formatted_content(msg.into())
            .to_owned()
    }

    pub fn new_system(msg: impl Into<String>) -> ChatMessage {
        ChatMessage::default()
            .with_role(ChatRole::System)
            .with_formatted_content(msg.into())
            .to_owned()
    }

    pub fn new_command(cmd: impl Into<String>) -> ChatMessage {
        ChatMessage::default()
            .with_role(ChatRole::Command)
            .with_formatted_content(cmd.into().to_string())
            .to_owned()
    }

    pub fn new_assistant(msg: impl Into<String>) -> ChatMessage {
        ChatMessage::default()
            .with_role(ChatRole::Assistant)
            .with_formatted_content(msg.into())
            .to_owned()
    }

    pub fn new_tool(msg: impl Into<String>) -> ChatMessage {
        ChatMessage::default()
            .with_role(ChatRole::Tool)
            .with_formatted_content(msg.into())
            .to_owned()
    }

    pub fn with_role(&mut self, role: ChatRole) -> &mut Self {
        self.role = role;
        self
    }

    pub fn with_formatted_content(&mut self, content: impl Into<String>) -> &mut Self {
        self.formatted_content = content.into();
        self
    }

    pub fn with_original(&mut self, original: swiftide::chat_completion::ChatMessage) -> &mut Self {
        self.original = Some(original);
        self
    }

    #[must_use]
    pub fn formatted_content(&self) -> &str {
        &self.formatted_content
    }
    #[must_use]
    pub fn role(&self) -> &ChatRole {
        &self.role
    }

    #[must_use]
    pub fn original(&self) -> Option<&swiftide::chat_completion::ChatMessage> {
        self.original.as_ref()
    }

    #[must_use]
    pub fn maybe_completed_tool_call(&self) -> Option<&swiftide::chat_completion::ToolCall> {
        match self.original() {
            Some(swiftide::chat_completion::ChatMessage::ToolOutput(tool_call, ..)) => {
                Some(tool_call)
            }
            _ => None,
        }
    }
}

impl From<swiftide::chat_completion::ChatMessage> for ChatMessage {
    fn from(msg: swiftide::chat_completion::ChatMessage) -> Self {
        let mut builder = match &msg {
            swiftide::chat_completion::ChatMessage::System(msg) => ChatMessage::new_system(msg),
            swiftide::chat_completion::ChatMessage::User(msg) => ChatMessage::new_user(msg),
            swiftide::chat_completion::ChatMessage::Assistant(msg, ..) => {
                ChatMessage::new_assistant(msg.as_deref().unwrap_or_default())
            }
            swiftide::chat_completion::ChatMessage::ToolOutput(tool_call, _) => {
                ChatMessage::new_tool(format!("tool `{}` completed", tool_call.name()))
            }
            swiftide::chat_completion::ChatMessage::Summary(_) => unimplemented!(),
        };

        builder.with_original(msg).to_owned()
    }
}
