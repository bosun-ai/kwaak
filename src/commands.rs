use std::sync::Arc;

use anyhow::Result;
use tokio::{sync::mpsc, task};
use uuid::Uuid;

use crate::{
    chat_message::ChatMessage,
    frontend::{App, UIEvent},
    indexing, query,
    repository::Repository,
};

/// Commands represent concrete actions from a user or in the backend
///
/// By default all commands can be triggered from the ui like `/<command>`
#[derive(
    Debug,
    PartialEq,
    Eq,
    strum_macros::EnumString,
    strum_macros::Display,
    strum_macros::IntoStaticStr,
    strum_macros::EnumIs,
    Clone,
)]
#[strum(serialize_all = "snake_case")]
pub enum Command {
    Quit {
        uuid: Uuid,
    },
    ShowConfig {
        uuid: Uuid,
    },
    IndexRepository {
        uuid: Uuid,
    },
    /// Default when no command is provided
    Chat {
        uuid: Uuid,
        message: String,
    },
}

impl Command {
    pub fn uuid(&self) -> Uuid {
        match self {
            Command::Quit { uuid }
            | Command::ShowConfig { uuid }
            | Command::IndexRepository { uuid }
            | Command::Chat { uuid, .. } => *uuid,
        }
    }

    pub fn with_uuid(self, uuid: Uuid) -> Self {
        match self {
            Command::Quit { .. } => Command::Quit { uuid },
            Command::ShowConfig { .. } => Command::ShowConfig { uuid },
            Command::IndexRepository { .. } => Command::IndexRepository { uuid },
            Command::Chat { message, .. } => Command::Chat { uuid, message },
        }
    }
}

/// Commands always flow via the `CommandHandler`
pub struct CommandHandler {
    /// Receives commands
    rx: mpsc::UnboundedReceiver<Command>,
    #[allow(dead_code)]
    /// Sends commands
    tx: mpsc::UnboundedSender<Command>,
    /// Sends `UIEvents` to the connected frontend
    ui_tx: Option<mpsc::UnboundedSender<UIEvent>>,
    /// Repository to interact with
    repository: Arc<Repository>,
}

impl CommandHandler {
    pub fn from_repository(repository: impl Into<Repository>) -> Self {
        let (tx, rx) = mpsc::unbounded_channel();

        CommandHandler {
            rx,
            tx,
            ui_tx: None,
            repository: Arc::new(repository.into()),
        }
    }

    pub fn register_ui(&mut self, app: &mut App) {
        self.ui_tx = Some(app.ui_tx.clone());
        app.command_tx = Some(self.tx.clone());
    }

    pub fn start(mut self) -> tokio::task::JoinHandle<()> {
        task::spawn(async move {
            while let Some(cmd) = self.rx.recv().await {
                let tx = self.ui_tx.clone().expect("Expected a registered ui");
                let repository = Arc::clone(&self.repository);

                tokio::spawn(async move {
                    if let Err(error) =
                        Self::handle_command(&Arc::clone(&repository), &tx, &cmd).await
                    {
                        tracing::error!(?error, %cmd, "Failed to handle command {cmd} with error {error:#}");
                        tx.send(
                            ChatMessage::new_system(format!("Failed to handle command: {error:#}"))
                                .uuid(cmd.uuid())
                                .to_owned()
                                .into(),
                        )
                        .unwrap();
                    }
                });
            }
        })
    }

    /// TODO: Most commands should probably be handled in a tokio task
    /// Maybe generalize tasks to make ui updates easier?
    #[tracing::instrument(skip(repository, tx))]
    async fn handle_command(
        repository: &Repository,
        tx: &mpsc::UnboundedSender<UIEvent>,
        cmd: &Command,
    ) -> Result<()> {
        let now = std::time::Instant::now();
        tracing::warn!("Handling command {cmd}");

        #[allow(clippy::match_wildcard_for_single_variants)]
        match cmd {
            Command::IndexRepository { .. } => indexing::index_repository(repository).await?,
            Command::ShowConfig { uuid } => {
                tx.send(
                    ChatMessage::new_system(toml::to_string_pretty(repository.config())?)
                        .uuid(*uuid)
                        .to_owned()
                        .into(),
                )
                .unwrap();
            }
            Command::Chat { uuid, ref message } => {
                let response = query::query(repository, message).await?;
                tracing::info!(%response, "Chat message received, sending to frontend");
                let response = ChatMessage::new_system(response).uuid(*uuid).to_owned();

                tx.send(response.into()).unwrap();
            }
            // Anything else we forward to the UI
            _ => tx.send(cmd.clone().into()).unwrap(),
        }
        let elapsed = now.elapsed();
        tx.send(
            ChatMessage::new_system(format!(
                "Command {cmd} successful in {} seconds",
                elapsed.as_secs_f64()
            ))
            .uuid(cmd.uuid())
            .into(),
        )
        .unwrap();

        Ok(())
    }
}
