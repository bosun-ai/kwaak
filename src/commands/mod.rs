//! Kwaak uses a command pattern to handle the backend asynchroniously.
mod command;
mod handler;
mod responder;

pub use command::{Command, CommandEvent};
pub use handler::CommandHandler;
pub use responder::{CommandResponse, Responder};

#[cfg(test)]
pub use responder::MockResponder;
