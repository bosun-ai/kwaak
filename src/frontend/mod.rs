mod app;
mod splash;
mod ui_event;
mod ui_input_command;

mod actions;
mod app_command_responder;
/// Different frontend ui modes
mod chat_mode;
mod logs_mode;

/// Let's be very strict about what to export
/// to avoid coupling frontend and the rest
pub use app::App;

#[cfg(debug_assertions)]
pub use app_command_responder::AppCommandResponder;
#[cfg(debug_assertions)]
pub use chat_mode::ui;
#[cfg(debug_assertions)]
pub use ui_event::UIEvent;
#[cfg(debug_assertions)]
pub use ui_input_command::{DiffVariant, UserInputCommand};
