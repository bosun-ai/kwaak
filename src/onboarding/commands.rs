use anyhow::Result;
use serde_json::json;

use crate::onboarding::util::prompt_text;

pub fn command_questions(context: &mut tera::Context) -> Result<()> {
    println!(
        "\nKwaak agents can run tests and use code coverage when coding. Kwaak uses tests as an extra feedback moment for agents"
    );

    let test_command =
        prompt_text("Test command (optional, <esc> to skip)", None).prompt_skippable()?;

    let coverage_command =
        prompt_text("Coverage command (optional, <esc> to skip)", None).prompt_skippable()?;

    context.insert(
        "commands",
        &json!({
            "test": test_command,
            "coverage": coverage_command,
        }),
    );

    Ok(())
}
