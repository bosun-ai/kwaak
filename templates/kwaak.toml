# This is a a prefilled template for `kwaak`
#
# Several assumptions and defaults have been filled in. For proper usage, please customize the values to your needs.
project_name = "{{ project_name }}"
language = "{{ language }}"

## If you are using OpenAI, set the api key here
#openai_api_key = "env:OPENAI_API_KEY"

## Optional: Connect kwaak to github to create PRs, search code, and automatically push to a  remote
#github_api_key = "env:GITHUB_TOKEN"

## Optional: Connect kwaak to tavily to enable it to search the web
#tavily_api_key = "env:TAVILY_API_KEY"

## Commands the agent uses for tools
## Test and coverage are required. Kwaak expects coverage to run the tests as well and echo both the test and coverage results to stdout
## At the moment, the format of the output does not matter.
[commands]
# Example: test = "cargo test --no-fail-fast --color=never"
test = "REQUIRED"
# Example: coverage = "cargo llvm-cov --no-clean --summary-only"
coverage = "REQUIRED"
## Optional: Lint and fix command. This command is run after each completion cycle, before committing the code.
## Recommended to use, as it avoids the LLM getting distracted by linting issues
# Example: lint_and_fix = "cargo clippy --fix --allow-dirty --allow-staged && cargo fmt"

## Git and GitHub configuration
#
## Kwaak can create and update PRs on Github, search github code, and interact with the git repository. This requires a github token.
## If you leave the token empty, kwaak will not create PRs.
[git]
owner = "{{ github.owner }}"
repository = "{{ github.repository }}"
main_branch = "{{ github.main_branch }}"
## The token is expected to have read and write access for git, and the ability to create and update pull requests.
# token = "env:KWAAK_GITHUB_TOKEN"

## Kwaak uses different LLMs for different tasks. As a rule of thumb, tasks that happen often (like indexing, summarizing) require a small, fast model
## and tasks that happen less often (like completion) can use a larger, more accurate model.
#
## You can overwrite the api key per kind of task if needed.
[llm.indexing]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
prompt_model = "gpt-4o-mini"

[llm.query]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
prompt_model = "gpt-4o"

[llm.embedding]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
embedding_model = "text-embedding-3-large"

## Docker configuration
## kwaak requires a Dockerfile for the tool execution environment.
## Besides the dependencies to run the code, there are several additional dependencies:
## - `git` for interacting with the codebase
## - `rg` (ripgrep) for searching the codebase
## - `fd` (fd) for effective file searching
##
## In the future, an executor is planned that does not have these dependencies, but for now, they are required.
##
## If your project already has a Dockerfile and you want to keep it clean, you can specify a different file to use.
[docker]
dockerfile = "Dockerfile"
