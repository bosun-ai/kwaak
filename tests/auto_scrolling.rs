use kwaak::chat_message::ChatMessage;
use kwaak::commands::Command;
use kwaak::frontend::{ui, UIEvent, UserInputCommand};
use kwaak::test_utils::{setup_integration, IntegrationContext};
use kwaak::{assert_agent_responded, assert_command_done};

#[test_log::test(tokio::test(flavor = "multi_thread"))]
async fn auto_scrolling() {
    let mut context = setup_integration().await.unwrap();

    for i in 0..100 {
        let event = UIEvent::ChatMessage(context.uuid, ChatMessage::new_user(format!("hello {i}")));
        context.app.send_ui_event(event);

        context.render_ui();

        context
            .app
            .handle_events_until(UIEvent::is_scroll_end)
            .await;
    }

    // Current chat records the number of rendred lines previously, so we need to call it twice
    insta::assert_snapshot!("auto_scrolled", context.render_ui());
}
