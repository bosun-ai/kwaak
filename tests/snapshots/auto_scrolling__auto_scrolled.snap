---
source: tests/auto_scrolling.rs
expression: context.render_ui()
---
"┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────↑─────────────Chats─────────────┐"
"│ ▶ hello 72                                                                                                                   ║ 󰍩  Chat #1                    │"
"│ ▶ hello 73                                                                                                                   ║                               │"
"│ ▶ hello 74                                                                                                                   ║                               │"
"│ ▶ hello 75                                                                                                                   ║                               │"
"│ ▶ hello 76                                                                                                                   ║                               │"
"│ ▶ hello 77                                                                                                                   ║                               │"
"│ ▶ hello 78                                                                                                                   ║                               │"
"│ ▶ hello 79                                                                                                                   ║                               │"
"│ ▶ hello 80                                                                                                                   ║                               │"
"│ ▶ hello 81                                                                                                                   ║                               │"
"│ ▶ hello 82                                                                                                                   ║                               │"
"│ ▶ hello 83                                                                                                                   ║                               │"
"│ ▶ hello 84                                                                                                                   ║                               │"
"│ ▶ hello 85                                                                                                                   ║                               │"
"│ ▶ hello 86                                                                                                                   ║                               │"
"│ ▶ hello 87                                                                                                                   ║                               │"
"│ ▶ hello 88                                                                                                                   █                               │"
"│ ▶ hello 89                                                                                                                   █─────────Chat commands─────────┤"
"│ ▶ hello 90                                                                                                                   █                               │"
"│ ▶ hello 91                                                                                                                   █  /new_chat       /help        │"
"│ ▶ hello 92                                                                                                                   █  /delete_chat    /quit        │"
"│ ▶ hello 93                                                                                                                   █  /copy           /show_config │"
"│ ▶ hello 94                                                                                                                   ║  /diff           /index_repos │"
"│ ▶ hello 95                                                                                                                   ║                               │"
"│ ▶ hello 96                                                                                                                   ║                               │"
"│ ▶ hello 97                                                                                                                   ║──────────Keybindings──────────┤"
"│ ▶ hello 98                                                                                                                   ║                               │"
"│ ▶ hello 99                                                                                                                   ║ Page Up/Down - Scroll         │"
"│                                                                                                                              ↓ End - Scroll to end           │"
"├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤ Tab - Next chat               │"
"│  Send a message to an agent ...                                                                                              │ ^s - Send message             │"
"│                                                                                                                              │ ^x - Stop agent               │"
"│                                                                                                                              │ ^n - New chat                 │"
"│                                                                                                                              │ ^q - Quit                     │"
"│                                                                                                                              │                               │"
"│                                                                                                                              │                               │"
"└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────────────────┘"
"                                                                                        uuid: a1a2a3a4-b1b2-c1c2-d1d2-d3d4d5d6d7d8   branch-name: not yet named "
"                                                                                                                                                                "
